File: backend_py/validation.py
from __future__ import annotations


def ensure_non_empty(value: str, field_name: str) -> None:
    if not value:
        raise ValueError(f"{field_name} must not be empty")


def ensure_max_bytes(value: str, field_name: str, max_bytes: int) -> None:
    if len(value.encode("utf-8")) > max_bytes:
        raise ValueError(f"{field_name} exceeds {max_bytes} bytes")

File: backend/validation.nim
import std/[strformat]

proc ensureNonEmpty*(value: string, fieldName: string) =
  if value.len == 0:
    raise newException(ValueError, fmt"{fieldName} must not be empty")

proc ensureMaxBytes*(value: string, fieldName: string, maxBytes: int) =
  if value.len > maxBytes:
    raise newException(ValueError, fmt"{fieldName} exceeds {maxBytes} bytes")
